libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -423,39 +432,61 @@ AP_GPS::update(void)
+        if (i == primary_instance) {
+            continue;
+        }
 
-            if (state[i].status == state[primary_instance].status && another_gps_has_1_or_more_sats) {
+        // work out which GPS is the primary
+        if (now - _last_instance_swap_ms < 20000 &&
+                state[primary_instance].status >= GPS_OK_FIX_3D) {
+            // don't allow switches too often except if primary is in trouble
+            continue;
+        }
 
-                uint32_t now = AP_HAL::millis();
-                bool another_gps_has_2_or_more_sats = (state[i].num_sats >= state[primary_instance].num_sats + 2);
+        switch (_auto_switch) {
+        default:
+        case GPS_AUTO_SWITCH_DISABLED:
+            break;
 
-                if ( (another_gps_has_1_or_more_sats && (now - _last_instance_swap_ms) >= 20000) ||
-                     (another_gps_has_2_or_more_sats && (now - _last_instance_swap_ms) >= 5000 ) ) {
-                // this GPS has more satellites than the
-                // current primary, switch primary. Once we switch we will
-                // then tend to stick to the new GPS as primary. We don't
-                // want to switch too often as it will look like a
-                // position shift to the controllers.
-                primary_instance = i;
-                _last_instance_swap_ms = now;
-                }
+        case GPS_AUTO_SWITCH_MORE_SATS:
+            if (state[i].status > state[primary_instance].status ||
+                (state[i].num_sats >= state[primary_instance].num_sats + 2)) {
+                // there is a higher status lock or 2+ sats
+                primary_instance_change_request = i;
             }
-        } else {
-            primary_instance = 0;
+            break;
+
+        case GPS_AUTO_SWITCH_VZ:
+            float pri = fabsf(state[primary_instance].velocity.z);
+            float sec = fabsf(state[i].velocity.z);
+            if (pri > _vz_switch_threshold && // above threshold
+                sec < _vz_switch_threshold && // below threshold
+                fabsf(pri - sec) > 0.1f && // diverged by >10%. A little bit of difference is normal so lets reject that
+                timing[primary_instance].last_fix_time_ms > 0 && // has locked before, keeps us from switching just after bootup
+                state[i].status >= GPS_OK_FIX_3D) { // sec is rdy
+                    primary_instance_change_request = i;
+                }
+            break;
+        } // switch
+    } // for i
+
+    // if primary index changed then note the time to inhibit doing this too often
+    if (primary_instance_change_request != primary_instance) {
+        // to protect against glitches, require the instance change request to persist
+        if (primary_instance_change_debounce_ms == 0) {
+            primary_instance_change_debounce_ms = now;
+        } else if (now - primary_instance_change_debounce_ms >= 500) {
```

Comment:
We'd need to touch base with @priseborough but I suspect this denounce delay is problematic for the EKF if we are waiting to swap because the primary has lost it's fix. (IE now we have a half second of drift + the error of the new primary GPS position).